### PR TITLE
Simplify DynamicWinsockMethods

### PR DESCRIPTION
**Pull request by jkotas**

### CODE REVIEW

**Review by geoffkizer — APPROVED** [review]

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/DynamicWinsockMethods.cs [hunk 1]
```diff
@@ -141,125 +79,27 @@ private IntPtr LoadDynamicFunctionPointer(SafeSocketHandle socketHandle, ref Gui
                 throw new SocketException();
             }
 
-            return ptr;
+            Interlocked.CompareExchange<T?>(ref cache, Marshal.GetDelegateForFunctionPointer<T>(ptr), null!);
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/DynamicWinsockMethods.cs:82]: I'm surprised the ! is necessary on the null.  What error do you get without that?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/DynamicWinsockMethods.cs [hunk 2]
```diff
@@ -38,86 +38,24 @@ public static DynamicWinsockMethods GetMethods(AddressFamily addressFamily, Sock
...
-                return (T)(object)_transmitPackets;
-            }
-
-            Debug.Fail("Invalid type passed to DynamicWinsockMethods.GetDelegate");
-            return null;
-        }
-
-        // Private methods that actually load the function pointers.
-        private IntPtr LoadDynamicFunctionPointer(SafeSocketHandle socketHandle, ref Guid guid)
+        private static T CreateDelegate<T>(ref T? cache, SafeSocketHandle socketHandle, string guidString) where T: Delegate
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/DynamicWinsockMethods.cs:56]: I think you should be able to make it:
```suggestion
        private static T CreateDelegate<T>([NotNull] ref T? cache, SafeSocketHandle socketHandle, string guidString) where T: Delegate
```

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Windows.cs [hunk 3]
```diff
@@ -151,12 +151,9 @@ public Socket EndAccept(out byte[] buffer, out int bytesTransferred, IAsyncResul
...
         }
 
-        private void EnsureDynamicWinsockMethods()
+        private DynamicWinsockMethods GetDynamicWinsockMethods()
         {
-            if (_dynamicWinsockMethods == null)
-            {
-                _dynamicWinsockMethods = DynamicWinsockMethods.GetMethods(_addressFamily, _socketType, _protocolType);
-            }
+            return _dynamicWinsockMethods ?? (_dynamicWinsockMethods = DynamicWinsockMethods.GetMethods(_addressFamily, _socketType, _protocolType));
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Windows.cs:156]:
```suggestion
            return _dynamicWinsockMethods ??= DynamicWinsockMethods.GetMethods(_addressFamily, _socketType, _protocolType);
```

**Review by stephentoub — APPROVED** [review]
Nice.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/DynamicWinsockMethods.cs (hunk 1 above)
Comment (jkotas) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/DynamicWinsockMethods.cs]: Left over from me trying to find the right combination of `!`s and `?`s to make the compiler happy...

**Pull request merged**